### ISSUE TITLE
Bert model instantiated from BertForMaskedLM.from_pretrained('bert-base-uncased') and BertForMaskedLM(BertConfig.from_pretrained('bert-base-uncased')) give different results 

### ISSUE THREAD

**Issue by christian-storm** [huggingface/transformers#954]
The two different methods for instantiating a model produce different losses.

`from pytorch_transformers import BertForMaskedLM, BertConfig, BertTokenizer
import torch

tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
input_ids = torch.tensor(tokenizer.encode("Hello, my dog is cute")).unsqueeze(0)  

config = BertConfig.from_pretrained('bert-base-uncased')
config_model = BertForMaskedLM(config)
config_model.eval()
with torch.no_grad():
    config_outputs = config_model(input_ids, masked_lm_labels=input_ids)
config_loss = config_outputs[0]
print(config_loss.item())

pretrained_model = BertForMaskedLM.from_pretrained('bert-base-uncased')
pretrained_model.eval()
with torch.no_grad():
    pretrained_outputs = pretrained_model(input_ids, masked_lm_labels=input_ids)
pretrained_loss = pretrained_outputs[0]
print(pretrained_loss.item())
assert config_loss.item() == pretrained_loss.item()`

The losses produced:
10.574708938598633
1.690806269645691

**Comment by yuchuang1979** [comment]
Hi, not only they give different results, but also BertModel(BertConfig.from_pretrained('bert-base-uncased')) will give different results each time you run it.

`import torch
from pytorch_transformers import BertTokenizer, BertModel, BertConfig, BertForNextSentencePrediction
import numpy as np

init_checkpoint_pt = "./bert-base-uncased/"

config = BertConfig.from_pretrained(init_checkpoint_pt)
config.output_hidden_states=True
print(config.output_hidden_states)
tokenizer = BertTokenizer.from_pretrained(init_checkpoint_pt)
model = BertModel.from_pretrained(init_checkpoint_pt,config=config)
model.eval()
input_ids = torch.tensor(tokenizer.encode("Hello, my dog is cute")).unsqueeze(0)  # Batch size 1
outputs = model(input_ids)

a = np.squeeze(outputs[0].detach().numpy())
avg = np.mean(a,axis = 0)`

**Comment by FeiWang96** [comment]
It seems that BertModel(config) returns a random intialized Bert model with the architecture as the config file indicates, because the __init__ function doesn't load pretrained weights. BertModel.from_pretrained() is the right function to load both model architecture and pretrained weights. It's the same for other bert classes.

**Comment by christian-storm** [comment]
Shouldn't BertConfig.from_pretrained('bert-base-uncased') return a config that loads pretrained weights instead of randomly initialized ones?  I thought was the whole point of the example code in the [docs](https://huggingface.co/pytorch-transformers/model_doc/bert.html#bertformaskedlm):


config = BertConfig.from_pretrained('bert-base-uncased')
tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')

model = BertForMaskedLM(config)
input_ids = torch.tensor(tokenizer.encode("Hello, my dog is cute")).unsqueeze(0)  # Batch size 1
outputs = model(input_ids, masked_lm_labels=input_ids)
loss, prediction_scores = outputs[:2]

**Comment by yuchuang1979** [comment]
It should return a config that loads pretrained models; however it does not act like this way.


Sent from Yahoo Mail for iPhone


On Sunday, August 4, 2019, 4:32 PM, Christian Storm <notifications@github.com> wrote:


Shouldn't BertConfig.from_pretrained('bert-base-uncased') return a config that loads pretrained weights instead of randomly initialized ones? I thought was the whole point of the example code in the docs:

config = BertConfig.from_pretrained('bert-base-uncased')
tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')

model = BertForMaskedLM(config)
input_ids = torch.tensor(tokenizer.encode("Hello, my dog is cute")).unsqueeze(0) # Batch size 1
outputs = model(input_ids, masked_lm_labels=input_ids)
loss, prediction_scores = outputs[:2]

**Comment by thomwolf** [comment]
You are right, the example in the doc is misleading.

The only way to load pretrained weights in a `model` is to call a `model_class.from_pretrained()` method. I'll fix the doc.

**Comment by FeiWang96** [comment]
I've fixed the examples of loading pretrained models in docstrings. #973
